영화 정보를 가져오는 API는 data 모듈에 구현

ParkChan opened the pull request:
요구사항
네이버 API를 사용해서 영화 정보를 가져와야 합니다

GET, v1/search/movie.json
프로그래밍 요구 사항
영화 정보를 가져오는 API는 data 모듈에 구현되어야 합니다
data 모듈의 구현체는 모두 internal 키워드가 붙어야 합니다

ParkChan committed:
[추가] 네이버 API 연동 및 테스트 코드 작성중

요구사항
네이버 API를 사용해서 영화 정보를 가져와야 합니다

GET, v1/search/movie.json
프로그래밍 요구 사항
영화 정보를 가져오는 API는 data 모듈에 구현되어야 합니다
data 모듈의 구현체는 모두 internal 키워드가 붙어야 합니다

ParkChan committed:
[추가] 네이버 API 연동 및 테스트 코드

BeokBeok commented on buildSrc/src/main/java/Dependencies.kt:
```diff
@@ -38,9 +41,10 @@ object TestConfig {
+    const val ASSERTJ_CORE = "org.assertj:assertj-core:3.21.0"
 
     const val MOCKK = "io.mockk:mockk:1.10.6"
+    //const val KOTLIN_TEST_JUNIT5 = "org.jetbrains.kotlin:kotlin-test-junit5:1.6.0-M1"
```
P2. 불필요한 주석은 제거해주세요

ParkChan replied:
429558b71489848816d67a32db154710d1f01bb5 임시 등록했던 항목으로 삭제하였습니다.

BeokBeok commented on buildSrc/src/main/java/Dependencies.kt:
```diff
@@ -85,11 +85,10 @@ object NetworkConfig {
+    const val LOGGING_INTERCEPTOR = "com.squareup.okhttp3:logging-interceptor:4.8.0"
     const val CONVERTER_MOSHI = "com.squareup.retrofit2:converter-moshi:$RETROFIT_VERSION"
-    const val LOGGING = "com.squareup.okhttp3:logging-interceptor:4.8.0"
+    const val GSON = "com.squareup.retrofit2:converter-gson:$RETROFIT_VERSION"
```
P2. Moshi를 쓰면 되서 Gson은 지워도 될 것 같아요

ParkChan replied:
4ae811deb17e595e43b2ddebd08ad88708888aa5 Gson 삭제하였습니다.

BeokBeok commented on data/src/main/java/com/chan/moviesearcher/data/RetrofitService.kt:
```diff
@@ -0,0 +1,60 @@
+            .client(client)
+            .build()
+
+    private fun movieService(): MovieApi =
+        retrofitBuild(
+            moshiConverter(
+                moshi(
+                    kotlinJsonAdapterFactory()
+                )
+            ),
+            okHttpClient(httpLoggingInterceptor())
+        ).create(MovieApi::class.java)
+
+    val movieApi: MovieApi by lazy {
+        movieService()
+    }
```
P2. 서비스를 생성하는 부분은 다른 클래스에서 구현하면 좋을 것 같아요

ParkChan replied:
0bef254e0adffe4e67dc0ca5775a8068a7a00048 서비스를 생성하는 부분 분리 하였습니다.

BeokBeok commented on data/src/main/java/com/chan/moviesearcher/data/source/MovieApi.kt:
```diff
@@ -0,0 +1,24 @@
+        @Query("display") display: Int = DEFAULT_DISPLAY_COUNT
+    ): MovieResponse
+
+    companion object {
+        const val DEFAULT_DISPLAY_COUNT = 10
+    }
```
P1. private 접근자를 붙여주세요

ParkChan replied:
ee757e6eb13b00b6f8f07697046ef839c7ed2261 private 접근자가 빠져있었네요. 추가하였습니다.

BeokBeok commented on data/src/main/java/com/chan/moviesearcher/data/source/MovieDataSourceImpl.kt:
```diff
@@ -0,0 +1,13 @@
+    private val movieApi: MovieApi
+) : MovieDataSource{
+    override suspend fun fetchMovies(page:Int, query: String): MovieResponse =
+        withContext(ioDispatcher){
+            movieApi.fetchMovies(page, query)
+        }
```
P3. 레트로핏 서비스 생성 시 suspend 함수로 만들면 내부적으로 io 디스패처에서 구동됩니다.

ParkChan replied:
b7cbdbb44894478022b7f908352982d2b991c24b 모르고 있던 부분이었는데 의견 감사합니다!😊

BeokBeok commented on data/src/resources/api-response/1.json:
```diff
@@ -0,0 +1,108 @@
+{
```
P2. 테스트 관련 json 파일은 `src/test/resources` 폴더에 위치시켜주세요

ParkChan replied:
26f6cc6ea60131bdf89338a5311d3099ffe514ef 리소스 위치는 `src/test/resources` 여기가 적합하겠네요! 수정하였습니다.

BeokBeok commented on data/src/test/kotlin/com/chan/moviesearcher/data/entity/MovieResponseTest.kt:
```diff
@@ -0,0 +1,156 @@
+        val json = File("src/resources/api-response/1.json").readText()
+        val response = jsonAdapter.fromJson(json) ?: MovieResponse()
+
+        assertThat(response.lastBuildDate).isEqualTo("Wed, 06 Oct 2021 22:57:59 +0900")
```
P1. Date 타입으로 받아서 처리해주세요

ParkChan replied:
오류 문제가 있어 우선적으로 String으로 그대로 두었습니다.

서버로부터 전달받은 Date 문자열은 아래와 같습니다.
```
Wed, 06 Oct 2021 22:57:59 +0900 
```

- 문제점
moshiAdapter를 사용하여 Date 타입으로 전달받도록 진행중에 문제가 발견되었습니다.

- 테스트 코드로는 아래와 같이 확인하여 보니 잘 동작 되는것으로 확인이 되네요. 신비로움...😂

[테스트코드]
```
val calendar = Calendar.getInstance()
calendar.time = Date("Fri, 08 Oct 2021 00:10:56 +0900") <= Depreccated 된 나쁜 코드입니다.하하
println(">>>>>>>>>>> ${calendar.time}")
```

[결과]
Fri Oct 08 00:10:56 KST 2021


[테스트 코드 적용사항]
```
private val moshi = Moshi.Builder()
        .add(Date::class.java, Rfc3339DateJsonAdapter().nullSafe())
        .add(KotlinJsonAdapterFactory())
        .build()
```
moshi adapter 사용을 위한 Gradle Dependencies 추가
```
"com.squareup.moshi:moshi-adapters:$MOSHI_VERSION"
```
Date로 받도록 수정
```
@Json(name = "lastBuildDate")
    val lastBuildDate: Date = Date(),  <= 문제 발생
```

BeokBeok replied:
Date Format을 API에서 내려주는 형식과 맞추니까 동작하네요

DateAdpater를 만들어서 fromJson 함수에 다음과 같은 값을 반환하게 해보세요
`SimpleDateFormat("EEE, dd MMM yyyy HH:mm:ss Z", Locale.US)`

---

```
    @Test
    fun `영화 정보 리스트에 대한 Json을 엔티티로 변환합니다`() {
        val response = jsonAdapter.fromJson(MOVIE_RESPONSE_JSON) ?: MovieResponse()

        assertThat(response.lastBuildDate.time).isEqualTo(1633528679000)
        assertThat(response.total).isEqualTo(88)
        assertThat(response.start).isEqualTo(2)
    }

```
![스크린샷 2021-10-08 오후 8 07 13](https://user-images.githubusercontent.com/48344355/136546224-c485e240-f884-41a3-b4d4-3e4a21e291b4.png)

ParkChan replied:
1d229f7934b0f7c40cd2f055a27aa01c3fbf110d, fa49c63375a4910cf6f5078e52358cb19e9f0769 도움 감사드립니다.

Date 파싱 변환에는 성공하였는데 
테스트 코드로 값을 equals로 비교 하고자하니 아래와 같은 문제가 있었습니다.

```
1.json안의 실제 String 값
👉Wed, 06 Oct 2021 22:57:59 +0900

Date로 전달 받은 값 String 출력
👉Wed Oct 06 22:57:59 KST 2021
```

BeokBeok replied:
데이터를 표현하는 포멧만 다를 뿐 실제 날짜는 동일합니다.
해서, 위에 예시로 보여드린 time으로 값을 비교하시거나 혹은 API에서 내려주는 포멧과 완전하게 일치하는 포멧을 찾으시면 될 것 같습니다 😅

BeokBeok commented on data/src/test/kotlin/com/chan/moviesearcher/data/source/MovieDataSourceImplTest.kt:
```diff
@@ -0,0 +1,29 @@
+
+    @Test
+    fun `영화 리스트 가져오기 기능 테스트`() = runBlocking {
+        movieDataSourceImpl.fetchMovies(1, "a")
+        coVerify {
+            movieApi.fetchMovies(1, "a")
+        }
```
P2. 단순히 함수 호출 여부를 물어보는 테스트보다는, 실제 Response를 모킹해서 테스트를 해보면 어떨까요?

BeokBeok replied:
753f42cf491fcbbfa3b6c8e15860d19e7fb4bbf3 실제 Response를 모킹해서 테스트 되도록 수정하였습니다.

BeokBeok reviewed (changes requested):
작업하시느라 고생이 많으십니다.
약간의 코맨트를 작성하였습니다. :smile:

<details><summary>코드리뷰 우선순위 규칙 (P1~P3)</summary>
P1 > ★★★ : 적극적으로 고민해주세요 (Request changes)
<br>
코드 수정이 반드시 필요하다고 판단되는 경우입니다.<br>
리뷰 요청자께서 만약 수용할 수 없는 상황이라면 적합한 의견을 들어 토론해주세요.
<br></br>
P2 > ★★☆ : 웬만하면 반영해 주세요 (Comment)
<br>
코드 수정을 권장하거나 리뷰 요청자께서 해당 코드를 작성하신 의도를 알고자 하는 경우입니다.<br>
리뷰 요청자께서는 가급적이면 이유를 들어 설명해주시거나 의견을 달아주세요.
<br></br>
P3 > ★☆☆ : 반영해도 좋고 넘어가도 좋습니다 (Approve)
<br>
해당 의견을 반영하는 게 좋을지 고민해 보는 정도면 충분합니다.<br>
만약 리뷰 요청자께서 리뷰어의 의견을 반영하셨더라도 리뷰어에게 별도로 리뷰 요청을 드리지 않아도 됩니다.
</details>

ParkChan committed:
[PR/수정] P2. 불필요한 주석은 제거해주세요

ParkChan committed:
[PR/수정] P2. Moshi를 쓰면 되서 Gson은 지워도 될 것 같아요

ParkChan committed:
[PR/수정] P2. 서비스를 생성하는 부분은 다른 클래스에서 구현하면 좋을 것 같아요

ParkChan committed:
[PR/수정] P1. private 접근자를 붙여주세요

ParkChan committed:
[PR/수정] P3. 레트로핏 서비스 생성 시 suspend 함수로 만들면 내부적으로 io 디스패처에서 구동됩니다.

ParkChan committed:
[PR/수정] P2. 테스트 관련 json 파일은 src/test/resources 폴더에 위치시켜주세요

ParkChan committed:
[PR/수정] P1. Date 타입으로 받아서 처리해주세요

문제)moshiAdapter를 사용하여 Date type으로 사용하려고 하였으나 문제 발생
확인)테스트 코드로 아래와 같이 체크하여 보면 잘 동작 되는것으로 확인

[테스트코드]
val calendar = Calendar.getInstance()
calendar.time = Date("Fri, 08 Oct 2021 00:10:56 +0900")
println(">>>>>>>>>>> ${calendar.time}")

[결과]
>>>>>>>>>>> Fri Oct 08 00:10:56 KST 2021

ParkChan committed:
[PR/수정] P2. 단순히 함수 호출 여부를 물어보는 테스트보다는, 실제 Response를 모킹해서 테스트를 해보면 어떨까요?

BeokBeok commented on app/src/test/java/com/chan/moviesearcher/ExampleUnitTest.kt:
```diff
@@ -1,8 +1,7 @@
-import org.junit.Test
-
-import org.junit.Assert.*
+import org.junit.jupiter.api.Assertions.assertEquals
```
P2. 불필요한 테스트 파일은 제거해도 될 것 같아요

ParkChan replied:
691166bac7a1512c13a8ae5c12da96a0a5c9469d 수정 과정에서 제거하였습니다~!

BeokBeok replied:
커밋 아이디와 리뷰 내용과 다른 것 같습니다.
ExampleUnitTest 파일 자체를 제거해도 될 것 같습니다.

ParkChan replied:
fa49c63375a4910cf6f5078e52358cb19e9f0769 파일자체를 제거했습니다.

BeokBeok commented on data/src/main/java/com/chan/moviesearcher/data/source/MovieDataSource.kt:
```diff
@@ -0,0 +1,8 @@
+
+internal interface MovieDataSource {
+
+    suspend fun fetchMovies(page: Int, query: String): Result<MovieResponse>
```
P3. Result는 ViewModel에서 접근하는 엔드포인트에 적용하는 것이 좋습니다

ParkChan replied:
430ae3bb98963d78963912d71e3fc0ca37939d8c, a63e06c0a6cdaa42c003fbf682a1197f774d326a 엔드포인트에서 적용하는것이 좋겠네요!

BeokBeok commented on data/src/test/kotlin/com/chan/moviesearcher/data/entity/MovieResponseTest.kt:
```diff
@@ -0,0 +1,160 @@
+        jsonAdapter = moshi.adapter(MovieResponse::class.java)
+    }
+
+    @Test
+    fun `영화 정보 리스트에 대한 Json을 엔티티로 변환합니다`() {
+        val response = jsonAdapter.fromJson(MOVIE_RESPONSE_JSON) ?: MovieResponse()
+
+        assertThat(response.total).isEqualTo(88)
+        assertThat(response.start).isEqualTo(2)
+    }
+
+    @Test
+    fun `json 파일을 읽어와서 엔티티로 변환합니다`() {
+
+        val json = File("src/test/resources/api-response/1.json").readText()
+        val response = jsonAdapter.fromJson(json) ?: MovieResponse()
+
+        assertThat(response.total).isEqualTo(88)
+        assertThat(response.start).isEqualTo(2)
+        assertThat(response.display).isEqualTo(10)
+    }
```
P3. 학습 목적이시겠지만, 둘 중 하나만 하셔도 됩니다 😄

ParkChan replied:
12839f3225df0e7feed6d6fd25e1831d9b67842f 학습목적으로 두가지 다 사용해보았습니다.😀 참조는 @BeokBeok 님의 RandomUser입니다. 하하핫
코드상 불필요하게 중복이고 테스트 코드가 길어진 관계로 String으로 받아오는 JSON 테스트 코드를 삭제하였습니다.

BeokBeok commented on data/src/test/kotlin/com/chan/moviesearcher/data/source/MovieDataSourceImplTest.kt:
```diff
@@ -0,0 +1,37 @@
+    internal fun `영화 리스트 가져오기 기능 테스트`() = runBlocking {
+        val mockResponse: MovieResponse = mockk()
+        coEvery {
+            movieApi.fetchMovies(1, "a")
```
P2. 함수 파라미터가 2개 이상인 경우에는 argument name을 써주시면 가독성이 좋아집니다 😄

ParkChan replied:
8870ad7fdaee344a793649341c5e306c0811b8bd 조언 감사드립니다. 수정하였습니다.

ParkChan committed:
[PR/수정] P3. Result는 ViewModel에서 접근하는 엔드포인트에 적용하는 것이 좋습니다.

ParkChan committed:
[PR/수정] P3. Result는 ViewModel에서 접근하는 엔드포인트에 적용하는 것이 좋습니다.

ParkChan committed:
[PR/수정] P3. 학습 목적이시겠지만, 둘 중 하나만 하셔도 됩니다 😄

ParkChan committed:
[PR/수정] P2. 함수 파라미터가 2개 이상인 경우에는 argument name을 써주시면 가독성이 좋아집니다 😄

ParkChan committed:
[PR/수정] P1. Date 타입으로 받아서 처리해주세요

ParkChan committed:
[PR/수정] P1. Date 타입으로 받아서 처리해주세요

ParkChan committed:
[PR/수정] P2. 불필요한 테스트 파일은 제거해도 될 것 같아요

BeokBeok commented on data/src/main/java/com/chan/moviesearcher/data/DateJsonAdapter.kt:
```diff
@@ -0,0 +1,39 @@
+import java.text.SimpleDateFormat
+import java.util.*
+
+class DateJsonAdapter : JsonAdapter<Date?>() {
```
P2. Date를 non-null로 처리할 수 있을 것 같아요

ParkChan replied:
aec0b5e2c7b1dc3c4112708487a5d4cbeea93276 non-null로 수정하였습니다.

BeokBeok commented on data/src/main/java/com/chan/moviesearcher/data/DateJsonAdapter.kt:
```diff
@@ -0,0 +1,39 @@
+        }
+    }
+
+    @Suppress("SimpleDateFormat")
```
P2. 불필요한 어노테이션 같아요

ParkChan replied:
aec0b5e2c7b1dc3c4112708487a5d4cbeea93276 불필요한 어노테이션 삭제하였습니다.

BeokBeok commented on data/src/main/java/com/chan/moviesearcher/data/entity/MovieResponse.kt:
```diff
@@ -0,0 +1,19 @@
+    @Json(name = "items")
+    val items: List<Item> = emptyList(),
+    @Json(name = "lastBuildDate")
+    val lastBuildDate: Date = Date(),
```
P1. 기본값을 현재 시간으로 적용하게 된다면, 해당 데이터가 기본값인지에 대한 판단은 어떻게 할 수 있을까요?

ParkChan replied:
f7d55ef7108cde7ccb8a7f16a20301abf6d17779 기본값이면 문제가 되겠네요! Nullable로 수정하였습니다.
```
val lastBuildDate: Date?,
```

BeokBeok replied:
nullable로 하셔도 되고 null이 정 싫으시면 Date(0) 으로 하셔도 됩니다

ParkChan replied:
의견 감사합니다. 0으로 설정해보았던 적은 없었는데 테스트 코드로 확인하였습니다.
```
val sdf = SimpleDateFormat("yyyy-MM-dd-HH:mm:ss")
sdf.timeZone = TimeZone.getTimeZone("UTC")
println(sdf.format(Date(0))) //1970-01-01-00:00:00
```

ParkChan committed:
[PR/수정] P2. 불필요한 어노테이션 같아요

ParkChan committed:
[PR/수정] P1. 기본값을 현재 시간으로 적용하게 된다면, 해당 데이터가 기본값인지에 대한 판단은 어떻게 할 수 있을까요?

BeokBeok commented on data/src/main/java/com/chan/moviesearcher/data/DateJsonAdapter.kt:
```diff
@@ -0,0 +1,38 @@
+
+    @Synchronized
+    @Throws(IOException::class)
+    override fun fromJson(reader: JsonReader): Date {
+        return if (reader.peek() == JsonReader.Token.NULL) {
+            reader.nextNull<Any>() as Date
+        } else {
+            val string = reader.nextString()
+            return dateFormat.parse(string)
+        }
+    }
```
P3. 이렇게 작성하셔도 됩니다

```
return try {
    dateFormat.parse(reader.nextString())
} catch (e: Exception) {
    Date(0)
}
```

ParkChan replied:
코드가 더 간결하고 보기 좋네요!

BeokBeok commented on data/src/main/java/com/chan/moviesearcher/data/DateJsonAdapter.kt:
```diff
@@ -0,0 +1,38 @@
+
+    @Synchronized
+    @Throws(IOException::class)
+    override fun toJson(writer: JsonWriter, value: Date?) {
```
P3. 여기도 nullable 제거해주세요
```suggestion
    override fun toJson(writer: JsonWriter, value: Date) {
```

ParkChan replied:
상위 클래스가 아래와 같아서 nullable을 제거 할 수 없습니다.
```
public abstract void toJson(JsonWriter writer, @Nullable T value) throws IOException;
```

BeokBeok reviewed (approved):
<img src="https://user-images.githubusercontent.com/48344355/108810482-8fca5400-75ee-11eb-9a24-9931a3877f9d.png" width="30%">
